### PR TITLE
Fix staking

### PR DESCRIPTION
**Pull request by pepoviola**
Add logic to support chains that use staking.

close https://github.com/paritytech/zombienet/issues/254

### CODE REVIEW

File: docs/src/network-definition-spec.md [hunk 1]
```diff
@@ -39,6 +39,8 @@ The network config can be provided both in `json` or `toml` format and each sect
   - `commandWithArgs`: (String) Override default command and args.
   - `args`: (Array of strings) Arguments to be passed to the `command`.
   - `validator`: (Boolean, default true) Pass the `--validator` flag to the `command`.
+  - `invulnerable`: (Boolean, default false) If true, the node will be added to `invulnerables` in the chain spec.
+  - `balance`: (number, default 2000000000000) Balance to set in `balances` for node's account.
```
Comment (wirednkod) [inline docs/src/network-definition-spec.md:43]: Lets add a TODO here - to make the numbers with less 0s if possible

File: src/chain-spec.ts [hunk 2]
```diff
@@ -36,22 +36,30 @@ function getAuthorityKeys(chainSpec: ChainSpec, keyType: KeyType = "session") {
...
-  if (!keys) return;
-
-  keys.length = 0;
-
-  if (keyType === "session") {
-    const runtime = getRuntimeConfig(chainSpec);
-    if (runtime.collatorSelection && runtime.collatorSelection.invulnerables)
-      runtime.collatorSelection.invulnerables.length = 0;
+  //clear keys
+  if (runtimeConfig && runtimeConfig.session)
```
Comment (wirednkod) [inline src/chain-spec.ts:48]:
```suggestion
  if (runtimeConfig?.session)
```

File: src/chain-spec.ts [hunk 3]
```diff
@@ -36,22 +36,30 @@ function getAuthorityKeys(chainSpec: ChainSpec, keyType: KeyType = "session") {
...
+    runtimeConfig.session.keys.length = 0;
+  // clear aura
+  if (runtimeConfig && runtimeConfig.aura)
```
Comment (wirednkod) [inline src/chain-spec.ts:51]:
```suggestion
  if (runtimeConfig?.aura)
```

File: src/chain-spec.ts [hunk 4]
```diff
@@ -36,22 +36,30 @@ function getAuthorityKeys(chainSpec: ChainSpec, keyType: KeyType = "session") {
...
+    runtimeConfig.aura.authorities.length = 0;
+
+  // clear collatorSelection
+  if (runtimeConfig && runtimeConfig.collatorSelection)
```
Comment (wirednkod) [inline src/chain-spec.ts:55]:
```suggestion
  if (runtimeConfig?.collatorSelection)
```

File: src/chain-spec.ts [hunk 5]
```diff
@@ -36,22 +36,30 @@ function getAuthorityKeys(chainSpec: ChainSpec, keyType: KeyType = "session") {
...
+    runtimeConfig.collatorSelection.invulnerables = [];
+
+  // Clear staking
+  if (runtimeConfig && runtimeConfig.staking) {
```
Comment (wirednkod) [inline src/chain-spec.ts:59]:
```suggestion
  if (runtimeConfig?.staking) {
```

File: src/chain-spec.ts [hunk 6]
```diff
@@ -60,15 +68,35 @@ export function clearAuthorities(
   );
 }
 
+export async function addBalances(specPath: string, nodes: Node[]) {
+  const chainSpec = readAndParseChainSpec(specPath);
+  const runtime = getRuntimeConfig(chainSpec);
+  for (const node of nodes) {
+    if (node.balance) {
```
Comment (wirednkod) [inline src/chain-spec.ts:75]:
```suggestion
    if (node?.balance) {
```

File: src/chain-spec.ts [hunk 7]
```diff
@@ -86,24 +114,41 @@ export async function addAuthority(
...
 
-  const chainSpec = readAndParseChainSpec(specPath);
-
   let keys = getAuthorityKeys(chainSpec);
   if (!keys) return;
 
   keys.push(key);
 
+  // staking
+  if (runtimeConfig.staking) {
```
Comment (wirednkod) [inline src/chain-spec.ts:123]:
```suggestion
  if (runtimeConfig?.staking) {
```

File: src/constants.ts [hunk 8]
```diff
@@ -80,6 +80,8 @@ export const DEV_ACCOUNTS = [
   "two",
 ];
 
+export const DEFAULT_BALANCE = 2000000000000;
```
Comment (wirednkod) [inline src/constants.ts:83]: Lets add a TODO here - to make the numbers with less 0s if possible

**Comment by pepoviola** [comment]
Thanks for the feedback @wirednkod! I made those changes :)

**Review by wirednkod — APPROVED** [review]

**Comment by wirednkod** [comment]
> Thanks for the feedback @wirednkod! I made those changes :)

Looks good. Thank you

**Pull request merged**